blockchain/event_filter.go

```diff
@@ -137,19 +152,20 @@ func (e *EventFilter) Events(cToken *ContinuationToken, chunkSize uint64) ([]*Fi
 			receipts = pending.Block.Receipts
 		}
 
-		var processedEvents uint64
 		matchedEvents, processedEvents, err = e.appendBlockEvents(matchedEvents, header, receipts, filterKeysMaps, cToken, chunkSize)
 		if err != nil {
 			if errors.Is(err, errChunkSizeReached) {
-				return matchedEvents, &ContinuationToken{
-					fromBlock:       curBlock,
-					processedEvents: processedEvents,
-				}, nil
+				rToken = &ContinuationToken{fromBlock: curBlock, processedEvents: processedEvents}
+				break
 			}
 			return nil, nil, err
 		}
 	}
-	return matchedEvents, nil, nil
+
+	if remainingScannedBlocks == 0 && curBlock <= e.toBlock {
+		rToken = &ContinuationToken{fromBlock: curBlock, processedEvents: processedEvents}
+	}
```

Comment:
Taking another look at this, I don't feel comfortable possibly overwriting the `rToken` set at L158 or using `processedEvents` here since it is returned by appendBlockEvents.

Let's make sure to only set `rToken` here only if it is `nil` and set 0 as its `processedEvents`